### PR TITLE
Reduce critical section for watchcache.lock

### PR DESCRIPTION
**Pull request by wojtek-t**
```release-note
NONE
```

/kind cleanup
/priority important-longterm
/sig api-machinery

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go [hunk 1]
```diff
@@ -312,25 +312,26 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
...
+		previousElem := previous.(*storeElement)
+		wcEvent.PrevObject = previousElem.Object
+		wcEvent.PrevObjLabels = previousElem.Labels
+		wcEvent.PrevObjFields = previousElem.Fields
+	}
+
 	if err := func() error {
-		// TODO: We should consider moving this lock below after the watchCacheEvent
-		// is created. In such situation, the only problematic scenario is Replace()
-		// happening after getting object from store and before acquiring a lock.
```
Comment (jpbetz) [inline staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:318]: Is this concern for Replace invalid? Or do we need to consider it?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go:318]: I no longer understand this comment (even though I was the one who added this).

But the story for replace is the following:
- we're using reflector and it's reflector that is calling the Replace
- the way reflector works is that it's effectively single-threaded - it's either processing watch events (and calling Add/Modify/Delete functions which land here) or calling Replace (and then this function can't be called).

So this safe.


The only interpretation that comes to my mind is that I had in mind some generic usage of watchcache, where the functions aren't called from reflector but by some arbitrary caller in arbitrary order from multiple threads. In such case this change wouldn't be safe, but it's not how it is used.

**Pull request merged**